src/notion.rs

```diff
@@ -164,21 +165,24 @@ See 'notion help <command>' for more information on a specific command.
         })
     }
 
-    fn run(self) -> Fallible<bool> {
+    fn run(self, session: &mut Session) -> Fallible<bool> {
         match self.command {
-            CommandName::Install => Install::go(self),
-            CommandName::Uninstall => Uninstall::go(self),
-            CommandName::Use => Use::go(self),
-            CommandName::Current => Current::go(self),
-            CommandName::Help => Help::go(self),
-            CommandName::Version => Version::go(self),
+            CommandName::Install => Install::go(self, session),
+            CommandName::Uninstall => Uninstall::go(self, session),
+            CommandName::Use => Use::go(self, session),
+            CommandName::Current => Current::go(self, session),
+            CommandName::Help => Help::go(self, session),
+            CommandName::Version => Version::go(self, session),
         }
     }
 }
 
 /// The entry point for the `notion` CLI.
 pub fn main() {
-    let exit_code = match Notion::go() {
+    let mut session = Session::new().unwrap();
```

Comment:
You don't want to `unwrap()` here, since it will turn the error into a panic, which will crash with an unhelpful Rust error message. Since you can't even generate events without a `Session`, you probably just need a different `match` for the `Session::new()` and the `Notion::go(...)` lines. But you can pull out the common error-handling logic of the two of them into a helper:

```rust
fn handle_errors<F: NotionFail>(err: &F) {
    if err.is_user_friendly() {
        display_error(&err);
    } else {
        display_unknown_error(&err);
    }

    if let Some(ref usage) = err.usage() {
        eprintln!();
        eprintln!("{}", usage);
    }
}

pub fn main() {
    let mut session = match Session::new() {
        Ok(session) => session,
        Err(err) => {
            handle_errors(&err);
            exit(1);
        }
    };

    session.add_event_start(ActivityKind::Notion);

    let exit_code = match Notion::go(&mut session) {
        ...
        Err(err) => {
            handle_errors(&err);
            err.exit_code()
        }
    };
    ...
}

Reply:
There's a similar issue in the Tool code that I cleaned up like this as well